rbc/heavydb_backend/buffer.py

```diff
@@ -343,36 +343,66 @@ def heavydb_buffer_getitem(x, i):
 
 # [rbc issue-197] Numba promotes operations like
 # int32(a) + int32(b) to int64
-def truncate_or_extend(builder, nb_value, eltype, value, buf_typ):
-    # buf[pos] = val
-
-    if isinstance(nb_value, types.Integer):  # Integer
-        if eltype.bitwidth < nb_value.bitwidth:
-            return builder.trunc(value, buf_typ)  # truncate
-        elif eltype.bitwidth > nb_value.bitwidth:
-            is_signed = nb_value.signed
-            return builder.sext(value, buf_typ) if is_signed else \
-                builder.zext(value, buf_typ)  # extend
-    elif isinstance(nb_value, types.Float):  # Floating-point
-        if eltype.bitwidth < nb_value.bitwidth:
-            return builder.fptrunc(value, buf_typ)  # truncate
-        elif eltype.bitwidth > nb_value.bitwidth:
-            return builder.fpext(value, buf_typ)  # extend
-    elif isinstance(nb_value, types.Boolean):
-        if buf_typ.width < value.type.width:
-            return builder.trunc(value, buf_typ)
-        elif buf_typ.width > value.type.width:
-            return builder.zext(value, buf_typ)
+def truncate_cast_or_extend(builder, value, typ, signed):
+
+    def _cast(builder, value, typ, signed):
+        fn = {
+            # unsigned int := float
+            (ir.IntType, ir.FloatType, False): builder.uitofp,
+            (ir.IntType, ir.DoubleType, False): builder.uitofp,
+            # signed int := float
+            (ir.IntType, ir.FloatType, True): builder.sitofp,
+            (ir.IntType, ir.DoubleType, True): builder.sitofp,
+
+            # float := unsigned int
+            (ir.FloatType, ir.IntType, False): builder.fptoui,
+            (ir.DoubleType, ir.IntType, False): builder.fptoui,
+            # float := signed int
+            (ir.FloatType, ir.IntType, True): builder.fptosi,
+            (ir.DoubleType, ir.IntType, True): builder.fptosi,
+        }[(value.type.__class__, typ.__class__, signed)]
```

Comment:
Nit that applies to all usages of `obj.__class__`:
```suggestion
        }[(type(value.type), type(typ), signed)]
```